flink/v1.18/flink/src/main/java/org/apache/iceberg/flink/util/FlinkVersionDetector.java

```diff
@@ -0,0 +1,44 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+package org.apache.iceberg.flink.util;
+
+import org.apache.flink.streaming.api.datastream.DataStream;
+
+public class FlinkVersionDetector {
+  public String version() {
+    String version = null;
+    try {
+      version = getVersionFromJar();
+    } catch (Exception e) {
+      /* we can't detect the exact implementation version from the jar (this can happen if the DataStream class
```

Comment:
Isn't it a packaging issue, if we have multiple `org.apache.flink.streaming.api.datastream.DataStream` classes on the classpath?

In this case the load order of these classes might depend on timining, and could cause issues.

Reply:
@pvary - Yes, it's a packaging issue. In my use case that reproduced the problem, the two instances of DataStream on the classpath are actually identical, due to unrelocated shading. That normally would cause no issues aside from build warnings. Of course we're fixing that too. 

Whether the packaging problem causes other bugs or not though, Iceberg's reaction shouldn't be "throw an NPE and crash the Flink pipeline", which is what currently happens and this PR fixes.

Reply:
Oh, and just fyi, I'm not affiliated with the engineers who posted the original issue this patch fixes; I just came across it while I was investigating. So this has happened at least twice in real-world use.

Reply:
Why did we get the `NullPointerException` in this case?
By my experience if we have the class on the classpath multiple times, Java just uses the first one it founds. Was it because the shading messed up with the metadata of the jar?

Reply:
I don't know for sure, but the shading messing up the metadata is my best guess too. 

In addition to the unit test in the patch, I did confirm that the patch resolves the issue in a real Flink cluster.